include/jemalloc/internal/guard.h

```diff
@@ -0,0 +1,76 @@
+#ifndef JEMALLOC_INTERNAL_GUARD_H
+#define JEMALLOC_INTERNAL_GUARD_H
+
+#include "jemalloc/internal/ehooks.h"
+#include "jemalloc/internal/emap.h"
+
+#define PAGE_GUARDS_SIZE (2 * PAGE)
+
+#define SAN_GUARD_LARGE_N_EXTENTS 0
+#define SAN_GUARD_SMALL_N_EXTENTS 0
+
+extern size_t opt_san_guard_large_nextents;
+extern size_t opt_san_guard_small_nextents;
+
+void guard_pages(tsdn_t *tsdn, ehooks_t *ehooks, edata_t *edata, emap_t *emap);
+void unguard_pages(tsdn_t *tsdn, ehooks_t *ehooks, edata_t *edata, emap_t *emap);
+void tsd_san_init(tsd_t *tsd);
+
+static inline bool
+san_enabled(void) {
+	return (opt_san_guard_large_nextents != 0 ||
+	    opt_san_guard_small_nextents != 0);
+}
+
+static inline bool
+large_extent_decide_guard(tsdn_t *tsdn, ehooks_t *ehooks, size_t size,
+    size_t alignment) {
+	if (opt_san_guard_large_nextents == 0 ||
+	    ehooks_guard_will_fail(ehooks) || tsdn_null(tsdn)) {
+		return false;
+	}
+
+	tsd_t *tsd = tsdn_tsd(tsdn);
+	uint64_t n = tsd_san_guard_large_nextent_get(tsd);
+	assert(n >= 1);
+	if (n > 1) {
+		/*
+		 * Subtract conditionally because the guard may not happen due
+		 * to alignment or size restriction below.
+		 */
+		*tsd_san_guard_large_nextentp_get(tsd) = n - 1;
+	}
+
+	if (n == 1 && (alignment <= PAGE) &&
+	    (size + PAGE_GUARDS_SIZE <= SC_LARGE_MAXCLASS)) {
+		*tsd_san_guard_large_nextentp_get(tsd) =
+		    opt_san_guard_large_nextents;
+		return true;
+	} else {
+		assert(tsd_san_guard_large_nextent_get(tsd) >= 1);
+		return false;
+	}
+}
+
+static inline bool
+slab_extent_decide_guard(tsdn_t *tsdn, ehooks_t *ehooks) {
```

Comment:
In general, is it as beneficial to guard slabs as large allocations?

For a large allocations I can imagine a large array with out of bounds access, which seems to be a common scenario. For slabs, most allocations, even if incorrectly accessed out of bounds, would much more ofter corrupt memory from the slab itself, rather than access guard pages.